From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Quintin Kong <2358543632@qq.com>
Date: Thu, 30 Jul 2026 19:32:47 -0400
Subject: [PATCH] 9438: #9434 Fix LoongArch byte and halfword reversal
 semantics

Several LoongArch reversal constructors use temporary/slice expressions that
duplicate source lanes in the generated pcode. The affected forms are revb.2h,
revb.4h, revb.2w, revb.d, revh.2w, and revh.d.

This replaces those expressions with explicit masks and shifts for each lane.

For input 0x0123456789abcdef, representative expected results are:

  revb.4h: 0x23016745ab89efcd
  revb.2w: 0x67452301efcdab89
  revb.d:  0xefcdab8967452301
  revh.2w: 0x45670123cdef89ab
  revh.d:  0xcdef89ab45670123

The modified language compiles with SLEIGH, and all six affected forms have been
checked against direct concrete reference functions.
---
 .../languages/loongarch32_instructions.sinc   |  5 +-
 .../languages/loongarch64_instructions.sinc   | 50 ++++++++-----------
 2 files changed, 22 insertions(+), 33 deletions(-)

diff --git a/Ghidra/Processors/Loongarch/data/languages/loongarch32_instructions.sinc b/Ghidra/Processors/Loongarch/data/languages/loongarch32_instructions.sinc
index cfc5ac21d2..48e66361d6 100644
--- a/Ghidra/Processors/Loongarch/data/languages/loongarch32_instructions.sinc
+++ b/Ghidra/Processors/Loongarch/data/languages/loongarch32_instructions.sinc
@@ -880,9 +880,8 @@
 #la-bitops-32.txt revb.2h mask=0x00003000	[@la32, @qemu]
 #0x00003000	0xfffffc00	r0:5,r5:5	['reg0_5_s0', 'reg5_5_s0']
 :revb.2h RD, RJ32src           is op10_31=0xc & RD & RJ32src {
-	tmp0:4 = (zext(RJ32src[0,8]) << 8) + zext(RJ32src[8,8]);
-	tmp1:4 = (zext(RJ32src[16,8]) << 8) + zext(RJ32src[24,8]);
-	RD = sext((tmp1 << 16) + tmp0);
+	local result:4 = ((RJ32src & 0x00ff00ff) << 8) | ((RJ32src & 0xff00ff00) >> 8);
+	RD = sext(result);
 }
 
 
diff --git a/Ghidra/Processors/Loongarch/data/languages/loongarch64_instructions.sinc b/Ghidra/Processors/Loongarch/data/languages/loongarch64_instructions.sinc
index 4e4442539e..086e8e728d 100644
--- a/Ghidra/Processors/Loongarch/data/languages/loongarch64_instructions.sinc
+++ b/Ghidra/Processors/Loongarch/data/languages/loongarch64_instructions.sinc
@@ -560,60 +560,50 @@
 #la-bitops-64.txt revb.4h mask=0x00003400	
 #0x00003400	0xfffffc00	r0:5,r5:5	['reg0_5_s0', 'reg5_5_s0']
 :revb.4h RD, RJsrc                    is op10_31=0xd & RD & RJsrc {
-	tmp0:8 = (zext(RJsrc[0,8])  << 8) + zext(RJsrc[8,8]);
-	tmp1:8 = (zext(RJsrc[16,8]) << 8) + zext(RJsrc[24,8]);
-	tmp2:8 = (zext(RJsrc[32,8]) << 8) + zext(RJsrc[40,8]);
-	tmp3:8 = (zext(RJsrc[48,8]) << 8) + zext(RJsrc[56,8]);
-
-	RD = (tmp3 << 48) + (tmp2 << 32) + (tmp1 << 16) + tmp0;
+	RD = ((RJsrc & 0x00ff00ff00ff00ff) << 8) |
+	     ((RJsrc & 0xff00ff00ff00ff00) >> 8);
 }
 
 
 #la-bitops-64.txt revb.2w mask=0x00003800	[@qemu]
 #0x00003800	0xfffffc00	r0:5,r5:5	['reg0_5_s0', 'reg5_5_s0']
 :revb.2w RD, RJsrc                    is op10_31=0xe & RD & RJsrc {
-	tmp0:8 = (zext(RJsrc[0,8])  << 24) + (zext(RJsrc[8,8])  << 16) + (zext(RJsrc[16,8]) << 8) + zext(RJsrc[24,8]);
-	tmp1:8 = (zext(RJsrc[32,8]) << 24) + (zext(RJsrc[40,8]) << 16) + (zext(RJsrc[48,8]) << 8) + zext(RJsrc[56,8]);
-
-	RD = (tmp1 << 32) + tmp0;
+	RD = ((RJsrc & 0x000000ff000000ff) << 24) |
+	     ((RJsrc & 0x0000ff000000ff00) << 8) |
+	     ((RJsrc & 0x00ff000000ff0000) >> 8) |
+	     ((RJsrc & 0xff000000ff000000) >> 24);
 }
 
 
 #la-bitops-64.txt revb.d mask=0x00003c00	[@qemu]
 #0x00003c00	0xfffffc00	r0:5,r5:5	['reg0_5_s0', 'reg5_5_s0']
 :revb.d RD, RJsrc                     is op10_31=0xf & RD & RJsrc {
-	tmp0:8 = zext(RJsrc[0,8]);
-	tmp1:8 = zext(RJsrc[8,8]);
-	tmp2:8 = zext(RJsrc[16,8]);
-	tmp3:8 = zext(RJsrc[24,8]);
-	tmp4:8 = zext(RJsrc[32,8]);
-	tmp5:8 = zext(RJsrc[40,8]);
-	tmp6:8 = zext(RJsrc[48,8]);
-	tmp7:8 = zext(RJsrc[56,8]);
-
-	RD = (tmp0 << 56) + (tmp1 << 48) + (tmp2 << 40) + (tmp3 << 32) + (tmp4 << 24) + (tmp5 << 16) + (tmp6 << 8) + tmp7;
+	RD = ((RJsrc & 0x00000000000000ff) << 56) |
+	     ((RJsrc & 0x000000000000ff00) << 40) |
+	     ((RJsrc & 0x0000000000ff0000) << 24) |
+	     ((RJsrc & 0x00000000ff000000) << 8) |
+	     ((RJsrc & 0x000000ff00000000) >> 8) |
+	     ((RJsrc & 0x0000ff0000000000) >> 24) |
+	     ((RJsrc & 0x00ff000000000000) >> 40) |
+	     ((RJsrc & 0xff00000000000000) >> 56);
 }
 
 
 #la-bitops-64.txt revh.2w mask=0x00004000	
 #0x00004000	0xfffffc00	r0:5,r5:5	['reg0_5_s0', 'reg5_5_s0']
 :revh.2w RD, RJsrc                    is op10_31=0x10 & RD & RJsrc {
-	tmp0:8 = (zext(RJsrc[0,16]) << 16) + zext(RJsrc[16,16]);
-	tmp1:8 = (zext(RJsrc[32,16]) << 8) + zext(RJsrc[48,16]);
-
-	RD = (tmp1 << 32) + tmp0;
+	RD = ((RJsrc & 0x0000ffff0000ffff) << 16) |
+	     ((RJsrc & 0xffff0000ffff0000) >> 16);
 }
 
 
 #la-bitops-64.txt revh.d mask=0x00004400	
 #0x00004400	0xfffffc00	r0:5,r5:5	['reg0_5_s0', 'reg5_5_s0']
 :revh.d RD, RJsrc                     is op10_31=0x11 & RD & RJsrc {
-	tmp0:8 = zext(RJsrc[0,16]);
-	tmp1:8 = zext(RJsrc[16,16]);
-	tmp2:8 = zext(RJsrc[32,16]);
-	tmp3:8 = zext(RJsrc[48,16]);
-
-	RD = (tmp3 << 48) + (tmp2 << 32) + (tmp1 << 16) + tmp0;
+	RD = ((RJsrc & 0x000000000000ffff) << 48) |
+	     ((RJsrc & 0x00000000ffff0000) << 16) |
+	     ((RJsrc & 0x0000ffff00000000) >> 16) |
+	     ((RJsrc & 0xffff000000000000) >> 48);
 }
 
 
-- 
2.45.1

